kernel/console/src/lib.rs

```diff
@@ -3,156 +3,166 @@
+        .name(format!("tty_to_{:?}", address))
+        .spawn()
+        .unwrap();
+    let writer_task = spawn::new_task_builder(port_to_tty_loop, (receiver, tty.master()))
+        .name(format!("{:?}_to_tty", address))
+        .spawn()
+        .unwrap();
+
+    let new_app_ns = mod_mgmt::create_application_namespace(None).unwrap();
+
+    // TODO: Replace shell with shell_2
+    let (app_file, _ns) =
+        mod_mgmt::CrateNamespace::get_crate_object_file_starting_with(&new_app_ns, "shell_2-")
+            .expect("Couldn't find shell in default app namespace");
+
+    let path = path::Path::new(app_file.lock().get_absolute_path());
+    let task = spawn::new_application_task_builder(path, Some(new_app_ns))
+        .unwrap()
+        .name(format!("{:?}_shell", address))
+        .block()
+        .spawn()
+        .unwrap();
+
+    let id = task.id;
+    let stream = Arc::new(tty.slave());
+    app_io::insert_child_streams(
+        id,
+        app_io::IoStreams {
+            discipline: Some(stream.discipline()),
+            stdin: stream.clone(),
+            stdout: stream.clone(),
+            stderr: stream,
+        },
+    );
+
+    task.unblock();
+    task.join().unwrap();
+
+    reader_task.kill(KillReason::Requested).unwrap();
+    writer_task.kill(KillReason::Requested).unwrap();
+
+    // Flush the tty in case the reader task didn't run between the last time the
+    // shell wrote something to the slave end and us killing the task.
+    let mut data = [0; 256];
+    if let Ok(len) = tty.master().try_read(&mut data) {
+        port.lock().write(&data[..len]).unwrap();
+    };
+
+    // TODO: Close port?
+}
+
+fn tty_to_port_loop((port, master): (Arc<MutexIrqSafe<SerialPort>>, tty::Master)) {
+    loop {
+        let mut data = [0; 256];
+        let len = master.read(&mut data).unwrap();
+        trace!("writing data to serial port: {:?}", &data[..len]);
+        port.lock().write(&data[..len]).unwrap();
+    }
+}
 
-/// The entry point for the each new [`Console`] task.
-fn console_entry<I, O, Backend>(
-	(mut console, input_receiver): (Console<I, O, Backend>, Receiver<DataChunk>),
-) -> Result<(), &'static str> 
-	where I: core2::io::Read,
-	      O: core2::io::Write,
-		  Backend: TerminalBackend,
-{
-	loop {
-		// Block until we receive the next data chunk from the sender.
-		match input_receiver.receive() {
-			Ok(DataChunk { len, data }) => {
-				let _res = console.terminal.handle_input(&mut &data[.. (len as usize)]);
-			}
-			Err(_e) => {
-				error!("[LIKELY BUG] Error receiving input data on {:?}: {:?}. Retrying...",
-					_e, console.name
-				);
-			}
-		}
-	}	
-
-	// Err("console_entry task returned unexpectedly")
+fn port_to_tty_loop((receiver, master): (Receiver<DataChunk>, tty::Master)) {
+    loop {
+        let DataChunk { data, len } = receiver.receive().unwrap();
+        trace!("read data from serial port: {:?}", &data[..len.into()]);
+        master.write(&data[..len as usize]).unwrap();
+    }
```

Comment:
Yes, the call to `read`/`receive` eventually boils down to `WaitQueue::wait_until`. We need separate tasks because there currently isn't a way to wait on two things at once.